clement-moulin-frier/vivarium
# Init state without configs

## Description


The PR modifies a lot of lines in the code, but there aren't tricky logic changes, we just modify the way to initialize the state without configs. I'll try to be as clear as I can when explaining the modifications.

I moved all the  state classes and definitions from ```sim_computation.py``` to a new file called ```states.py```.

The modifications also include refactoring imports in multiple files to align with the new file structure, enhancing readability and preventing linting errors.

In states.py, five functions have been implemented to initialize various states, such as simulator_state, nve_state, agent_state, objects_state, and the combined state. These changes offer greater flexibility in initializing states, allowing for customization such as specifying agent positions or using random values. Future iterations will extend this approach to more attributes, such as colors, diameters, behaviors ... to provide even more customization options.

Additionally, default values have been defined within these classes, and type hints have been added to functions for clarity. But I can still add some more docstrings to define what each parameter does.


## Related Issue (if applicable)

closes #51 

## How to Test
Launch the server
```
python3 scripts/run_server.py
```
Launch the Panel interface
```
panel serve scripts/run_interface.py --autoreload
```


Check if nothing is broken with these two lines, normally it should be the case because only code changing the state init has been modified.

Normally I added some ugly lines to show that you can initiate the position of agents with a list in run_server. Run :

```
python3 scripts/run_server.py --custom_pos
```

You can also try to launch the simulation with a custom number of agents or objects (will be usefull later for eco-evo simulations). Actually for this feature I think it would make more sense to rename the current ```n_agents``` to ```max_agents```, and ```n_existing_agents``` to ```n_agents``` (same for objects).

```
python3 scripts/run_server.py --n_existing_agents 7 --n_existing_objects 1
```

And relaunch the interface, check that the agents are indeed defined on the diagonal of the box. Later, this will easiely enable to define custom scenes, notably when using external files to store th epositions, colors ... as mentionned in #54 

## Screenshots (if applicable)
